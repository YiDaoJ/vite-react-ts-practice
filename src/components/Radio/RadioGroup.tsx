import { css } from "@emotion/react";
import React, { FC, PropsWithChildren } from "react";

interface RadioGroupProps {
  label?: string;
}

export const RadioGroup: FC<PropsWithChildren<RadioGroupProps>> = ({
  label,
  children,
}) => {
  return (
    <div className="radio-group__container" data-testid="radio-group">
      {label && (
        <label className="radio-group__label" css={GroupLabelStyle}>
          {label}
        </label>
      )}
      <div className="radio-group__control">{children}</div>
    </div>
  );
};

const GroupLabelStyle = css`
  display: block;
  font-size: 1rem;
  line-height: 1.5rem;
  letter-spacing: 0.011875rem;
  margin-bottom: 0.5rem;
  color: #6c767e;
`;
