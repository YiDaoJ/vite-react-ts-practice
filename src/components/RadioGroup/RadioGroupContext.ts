import React from "react";

export interface RadioGroupContextValue {
  name: string | undefined;
  onChange: (
    event: React.ChangeEvent<HTMLInputElement>,
    value?: string
  ) => void;
  selected: any;
}

const RadioGroupContext = React.createContext<
  RadioGroupContextValue | undefined
>(undefined);

export default RadioGroupContext;
