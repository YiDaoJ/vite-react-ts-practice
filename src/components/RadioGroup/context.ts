import React from "react";

export interface RadioGroupContextValue {
  name: string | undefined;
  selected: string;
  onChange: (
    event: React.ChangeEvent<HTMLInputElement>,
    value?: string
  ) => void;
}

const RadioGroupContext = React.createContext<
  RadioGroupContextValue | undefined
>(undefined);

export default RadioGroupContext;
