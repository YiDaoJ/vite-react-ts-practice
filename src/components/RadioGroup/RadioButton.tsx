import { css } from "@emotion/react";
import styled from "@emotion/styled";
import React, { ChangeEvent, FC, useContext, useEffect, useState } from "react";
import RadioGroupContext from "./context";

interface RadioButtonProps extends RadioButtonStyleProps {
  label: string;
  value: string;
}

interface RadioButtonStyleProps {
  disabled?: boolean;
}

export const RadioButton: FC<RadioButtonProps> = ({
  label,
  value,
  disabled,
}) => {
  const { selected, onChange } = useContext(RadioGroupContext);

  return (
    <RadioWrapper disabled={disabled}>
      <span className='radio-input'>
        <StyledInput
          type='radio'
          name={value}
          onChange={onChange}
          value={value}
          checked={selected === value}
          disabled={disabled}
        />
      </span>
      <span className='radio-label' css={labelStyle}>
        {label}
      </span>
    </RadioWrapper>
  );
};

// ------ wrapper -------
const RadioWrapper = styled.label<RadioButtonStyleProps>`
  display: inline-flex;
  margin-bottom: 1rem;
  position: relative;

  ${({ disabled }) => disabled && wrapperDisabled};
`;

const wrapperDisabled = css`
  color: #6c767e;
  cursor: not-allowed;
`;

// ------ input -------
const StyledInput = styled.input`
  appearance: none !important;
  margin-bottom: 0.3125rem;
  border-color: transparent;
  margin-top: 0.7rem;

  border-radius: 0.3125rem;
  display: inline-block;
  position: relative;
  padding-left: 2rem;
  outline: none;
  height: 1px;

  ::before {
    content: "";
    position: absolute;
    top: 50%;
    left: 0;
    transform: translate(-20%, -50%);
    width: 1.3125rem;
    height: 1.3125rem;
    border: 0.0625rem solid #6c767e;
    box-shadow: 0.0625rem 0.0625rem transparent,
      -0.0625rem -0.0625rem transparent, 0.0625rem -0.0625rem transparent,
      -0.0625rem 0.0625rem transparent;
    border-radius: 100%;
  }

  ${({ disabled }) => (disabled ? inputDisabled : inputActive)};
`;

const inputDisabled = css`
  pointer-events: none;
`;

const inputActive = css`
  cursor: pointer;
  :focus::before {
    border: 0.0625rem solid #000;
    outline: none;
    box-shadow: 0 0 0 0.3125rem rgb(108 118 126 / 24%);
  }

  :hover::before {
    border: 0.0625rem solid #000;
  }

  ::after {
    position: absolute;
    top: 50%;
    left: 0;
    transform: translate(-20%, -50%);
    content: "";
  }

  :checked {
    ::after {
      position: absolute;
      top: 50%;
      left: 0;
      display: block;
      border: 0.375rem solid #e00;
      border-radius: 100%;
      transform-origin: 0 0;
      transform: translate(10%, -50%);
      content: "";
    }
  }
`;

// ------ label -------
const labelStyle = css`
  font-size: 1rem;
  line-height: 1.5rem;
  letter-spacing: 0.011875rem;
  font-family: Verdana, Arial, sans-serif;
`;
