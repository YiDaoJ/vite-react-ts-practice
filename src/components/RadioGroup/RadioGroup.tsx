import { css } from "@emotion/react";
import React, { ChangeEvent, FC, PropsWithChildren, useState } from "react";
import RadioGroupContext from "./context";

interface RadioGroupStyleProps {
  horizontal?: boolean;
}

interface RadioGroupProps extends RadioGroupStyleProps {
  label?: string;
  name?: string;
  defaultValue?: string;
  onChange?(value: string): void;
}

export const RadioGroup: FC<PropsWithChildren<RadioGroupProps>> = ({
  label,
  name,
  defaultValue,
  onChange,
  horizontal,
  children,
}) => {
  const [value, setValue] = useState(defaultValue);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setValue(event.currentTarget.value);
    if (onChange) {
      onChange(event.currentTarget.value);
    }
  };

  return (
    <div className='radio-group__container' data-testid='radio-group'>
      {label && (
        <label className='radio-group__label' css={GroupLabelStyle}>
          {label}
        </label>
      )}
      <RadioGroupContext.Provider
        value={{ name, onChange: handleChange, selected: value }}
      >
        <div
          className='radio-group__control'
          css={horizontal ? GroupControlHorizontal : GroupControlVertical}
        >
          {children}
        </div>
      </RadioGroupContext.Provider>
    </div>
  );
};

const GroupLabelStyle = css`
  display: block;
  font-size: 1rem;
  line-height: 1.5rem;
  letter-spacing: 0.011875rem;
  margin-bottom: 0.5rem;
  color: #6c767e;
  font-family: Verdana, Arial, sans-serif;
`;

const GroupControlHorizontal = css`
  display: flex;
  flex-direction: row;
  gap: 2rem;
  justify-content: flex-start;
`;

const GroupControlVertical = css`
  display: flex;
  flex-direction: column;
`;
