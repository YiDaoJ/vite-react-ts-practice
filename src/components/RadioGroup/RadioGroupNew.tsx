import { css } from "@emotion/react";
import React, { ChangeEvent, FC, PropsWithChildren, useState } from "react";
import RadioGroupContext from "./RadioGroupContext";

interface RadioGroupProps {
  label?: string;
  name?: string;
  defaultValue?: string;
  onChange(value: string): void;
}

export const RadioGroup: FC<PropsWithChildren<RadioGroupProps>> = ({
  label,
  name,
  defaultValue,
  onChange,
  children,
}) => {
  const [value, setValue] = useState(defaultValue);

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setValue(event.currentTarget.value);
    if (onChange) {
      onChange(event.currentTarget.value);
    }
  };

  return (
    <div className='radio-group__container' data-testid='radio-group'>
      {label && (
        <label className='radio-group__label' css={GroupLabelStyle}>
          {label}
        </label>
      )}
      <RadioGroupContext.Provider
        value={{ name, onChange: handleChange, selected: value }}
      >
        <div className='radio-group__control'>{children}</div>
      </RadioGroupContext.Provider>
    </div>
  );
};

const GroupLabelStyle = css`
  display: block;
  font-size: 1rem;
  line-height: 1.5rem;
  letter-spacing: 0.011875rem;
  margin-bottom: 0.5rem;
  color: #6c767e;
  font-family: Verdana, Arial, sans-serif;
`;
