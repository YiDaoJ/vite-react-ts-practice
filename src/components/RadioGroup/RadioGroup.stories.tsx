import React from "react";
import { ComponentMeta, ComponentStory } from "@storybook/react";
import { RadioGroup } from "./RadioGroup.tsx";
import { RadioButton } from "./RadioButton";

export default {
  title: "Component/Radio Group",
  component: RadioGroup,
} as ComponentMeta<typeof RadioGroup>;

const Template: ComponentStory<typeof RadioGroup> = (args) => (
  <RadioGroup {...args}>
    <RadioButton value='cable' label='Kabel' />
    <RadioButton value='wifi' label='WI-FI' />
  </RadioGroup>
);

export const Standard = Template.bind({});

Standard.args = { label: "ANSCHLUSSART", defaultValue: "wifi" };
