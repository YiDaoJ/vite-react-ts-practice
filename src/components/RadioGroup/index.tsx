import React, { FC } from "react";
import { RadioGroup } from "./RadioGroupNew";
import { RadioButton } from "./RadioButtonNew";

export const RadioGroupRebuild: FC = () => {
  const handleChange = (value: string) => {
    const info = { selected: value };

    fetch("http://localhost:4000/test", {
      method: "POST",
      // mode: "no-cors",
      headers: {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Credentials": true,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(info),
    })
      .then((res) => res.json())
      .then((data) => console.log(data));
  };

  return (
    <RadioGroup
      label='ANSCHLUSSART'
      defaultValue='wifi'
      onChange={handleChange}
    >
      <RadioButton value='cable' label='Kabel' />
      <RadioButton value='wifi' label='WI-FI' />
    </RadioGroup>
  );
};
