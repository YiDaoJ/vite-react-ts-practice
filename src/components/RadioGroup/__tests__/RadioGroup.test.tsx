import { render, screen, fireEvent } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import { RadioButton } from "../RadioButton";
import { RadioGroup } from "../RadioGroup";

describe("RadioGroup test", () => {
  test("render RadioGroup component", () => {
    const { asFragment } = render(
      <RadioGroup label='Gender' defaultValue='female'>
        <RadioButton value='female' label='Female' />
        <RadioButton value='male' label='Male' />
      </RadioGroup>
    );
    // screen.debug(asFragment());
    expect(asFragment()).toMatchSnapshot();
  });

  test("RadioGroup change event", async () => {
    const handleChange = jest.fn();
    const user = userEvent.setup();

    render(
      <RadioGroup label='Gender' defaultValue='female'>
        <RadioButton value='female' label='Female' />
        <RadioButton value='male' label='Male' />
      </RadioGroup>
    );

    const element = screen.getByDisplayValue("male");
    await user.click(element);
    expect(element).toBeChecked();
  });
});
