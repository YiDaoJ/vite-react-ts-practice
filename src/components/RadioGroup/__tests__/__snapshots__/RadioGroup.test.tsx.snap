// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup test render RadioGroup component 1`] = `
<DocumentFragment>
  <div
    class="radio-group__container"
    data-testid="radio-group"
  >
    <label
      class="radio-group__label css-6sc5xn-GroupLabelStyle"
    >
      Gender
    </label>
    <div
      class="radio-group__control css-19gs4bv-GroupControlVertical"
    >
      <label
        class="css-5rvwsk-RadioWrapper e1jw82t61"
      >
        <span
          class="radio-input"
        >
          <input
            checked=""
            class="css-1ayomz5-StyledInput-inputActive e1jw82t60"
            name="female"
            type="radio"
            value="female"
          />
        </span>
        <span
          class="radio-label css-1wlzwar-labelStyle"
        >
          Female
        </span>
      </label>
      <label
        class="css-5rvwsk-RadioWrapper e1jw82t61"
      >
        <span
          class="radio-input"
        >
          <input
            class="css-1ayomz5-StyledInput-inputActive e1jw82t60"
            name="male"
            type="radio"
            value="male"
          />
        </span>
        <span
          class="radio-label css-1wlzwar-labelStyle"
        >
          Male
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;
